File: 150.evaluate-reverse-polish-notation.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=150 lang=cpp
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
class Solution
{
public:
    int evalRPN(vector<string> &tokens)
    {
        stack<int> stk;
        long a, b;
        for (string &token : tokens)
        {
            if (token == "+")
            {
                a = stk.top();
                stk.pop();
                b = stk.top();
                stk.pop();
                stk.push(b + a);
            }
            else if (token == "-")
            {
                a = stk.top();
                stk.pop();
                b = stk.top();
                stk.pop();
                stk.push(b - a);
            }
            else if (token == "*")
            {
                a = stk.top();
                stk.pop();
                b = stk.top();
                stk.pop();
                stk.push(b * a);
            }
            else if (token == "/")
            {
                a = stk.top();
                stk.pop();
                b = stk.top();
                stk.pop();
                stk.push(b / a);
            }
            else
            {
                stk.push(stoi(token));
            }
        }
        return stk.top();
    }
};
// @lc code=end
// int main()
// {
//     Solution sol;
//     vector<string> tokens = {"4", "13", "5", "/", "+"};
//     cout << sol.evalRPN(tokens);
//     return 0;
// }

File: Java/Grind75/Week 3/150.evaluate-reverse-polish-notation.java
import java.util.*;

/*
 * @lc app=leetcode id=150 lang=java
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stk = new Stack<>();
        int a = 0, b = 0;
        for (String str : tokens) {
            switch (str) {
                case "+":
                    a = stk.pop();
                    b = stk.pop();
                    stk.push(b
                            + a);
                    break;
                case "-":
                    a = stk.pop();
                    b = stk.pop();
                    stk.push(b
                            - a);
                    break;
                case "*":
                    a = stk.pop();
                    b = stk.pop();
                    stk.push(b
                            * a);
                    break;
                case "/":
                    a = stk.pop();
                    b = stk.pop();
                    stk.push(b
                            / a);
                    break;
                default:
                    stk.push(Integer.parseInt(str));
                    break;
            }
        }
        return (stk.peek());
    }
}
// @lc code=end
